File: backend/app/detectors/statistics.py
import re
from typing import List
from ..models.schemas import TextStatistics


def extract_words(text: str) -> List[str]:
    matches = re.findall(r"\b[a-zA-Z0-9'-]+\b", text)
    return [w for w in matches if re.search(r"[a-zA-Z0-9]", w)]


def extract_sentences(text: str) -> List[str]:
    if not text.strip():
        return []

    # Protect abbreviations and decimals
    protected = re.sub(
        r"\b(Dr|Mr|Mrs|Ms|Prof|Sr|Jr|vs|etc|e\.g|i\.e|U\.S|Jan|Feb|Mar|Apr|Aug|Sept|Oct|Nov|Dec)\.",
        r"\1__DOT__",
        text,
        flags=re.IGNORECASE,
    )
    protected = re.sub(r"(\d+)\.(\d+)", r"\1__DEC__\2", protected)

    raw_splits = re.split(r"(?<=[.!?])\s+(?=[A-Z0-9\"“'‘(])", protected)

    sentences = []
    for s in raw_splits:
        restored = s.replace("__DOT__", ".").replace("__DEC__", ".").strip()
        if restored:
            sentences.push(restored) if hasattr(sentences, "push") else sentences.append(restored)

    if not sentences and text.strip():
        return [text.strip()]

    return sentences


def extract_paragraphs(text: str) -> List[str]:
    if not text.strip():
        return []
    return [p.strip() for p in re.split(r"\n\s*\n", text) if p.strip()]


def calculate_statistics(text: str) -> TextStatistics:
    char_count = len(text)
    words = extract_words(text)
    word_count = len(words)
    sentences = extract_sentences(text)
    sentence_count = len(sentences)
    paragraphs = extract_paragraphs(text)
    paragraph_count = len(paragraphs)

    total_word_chars = sum(len(re.sub(r"['-]", "", w)) for w in words)
    avg_word_length = round(total_word_chars / word_count, 1) if word_count > 0 else 0.0
    avg_sentence_length = round(word_count / sentence_count, 1) if sentence_count > 0 else 0.0

    return TextStatistics(
        charCount=char_count,
        wordCount=word_count,
        sentenceCount=sentence_count,
        paragraphCount=paragraph_count,
        avgWordLength=avg_word_length,
        avgSentenceLength=avg_sentence_length,
    )

File: src/engine/statistics.ts
import { TextStatistics } from '../types/detector.ts';

/**
 * Split text into tokens and sentences using deterministic regex heuristics.
 */
export function extractWords(text: string): string[] {
  const matches = text.match(/\b[a-zA-Z0-9'-]+\b/g);
  return matches ? matches.filter((w) => /[a-zA-Z0-9]/.test(w)) : [];
}

/**
 * Split text into sentences, protecting common abbreviations and decimal numbers.
 */
export function extractSentences(text: string): string[] {
  if (!text.trim()) return [];

  // Temporarily mask common abbreviations and decimals to prevent erroneous sentence splits
  const protectedText = text
    .replace(/\b(Dr|Mr|Mrs|Ms|Prof|Sr|Jr|vs|etc|e\.g|i\.e|U\.S|Jan|Feb|Mar|Apr|Aug|Sept|Oct|Nov|Dec)\./gi, '$1__DOT__')
    .replace(/(\d+)\.(\d+)/g, '$1__DEC__$2');

  const rawSplits = protectedText.split(/(?<=[.!?])\s+(?=[A-Z0-9"“'‘(])/);

  const sentences: string[] = [];
  for (const s of rawSplits) {
    const restored = s
      .replace(/__DOT__/g, '.')
      .replace(/__DEC__/g, '.')
      .trim();
    if (restored.length > 0) {
      sentences.push(restored);
    }
  }

  // Fallback if regex resulted in no sentences but text is present
  if (sentences.length === 0 && text.trim().length > 0) {
    return [text.trim()];
  }

  return sentences;
}

/**
 * Extract non-empty paragraphs.
 */
export function extractParagraphs(text: string): string[] {
  if (!text.trim()) return [];
  return text
    .split(/\n\s*\n/)
    .map((p) => p.trim())
    .filter((p) => p.length > 0);
}

/**
 * Compute Step 1 Basic Text Statistics.
 */
export function calculateStatistics(text: string): TextStatistics {
  const charCount = text.length;
  const words = extractWords(text);
  const wordCount = words.length;
  const sentences = extractSentences(text);
  const sentenceCount = sentences.length;
  const paragraphs = extractParagraphs(text);
  const paragraphCount = paragraphs.length;

  const totalWordChars = words.reduce((acc, w) => acc + w.replace(/['-]/g, '').length, 0);
  const avgWordLength = wordCount > 0 ? Number((totalWordChars / wordCount).toFixed(1)) : 0;
  const avgSentenceLength = sentenceCount > 0 ? Number((wordCount / sentenceCount).toFixed(1)) : 0;

  return {
    charCount,
    wordCount,
    sentenceCount,
    paragraphCount,
    avgWordLength,
    avgSentenceLength,
  };
}
